src/aarch64/ffi.c

```diff
@@ -728,34 +737,95 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
   else if (flags & AARCH64_RET_NEED_COPY)
     rsize = 16;
 
-  /* Allocate consectutive stack for everything we'll need.
-     The frame uses 40/80 bytes for: lr, fp, rvalue, flags, sp */
-  context = alloca (sizeof(struct call_context) + stack_bytes + sizeof(struct call_frame) + rsize);
+  /*
+   * Allocate consectutive stack for everything we'll need.
+   * The frame uses 40/80 bytes for: lr, fp, rvalue, flags, sp.
+   *
+   * On CHERI-extended architectures, the allocation uses gaps between
+   * sub-objects that have the size equal to the alignment of a sub-object minus
+   * one. The one byte optimisation can be * introduced as for each sub-object
```

Comment:
* in the middle of the line

Reply:
Also, technically you can do better than this by using the known minimum alignment for the end of the previous subobject. But also all this really needs is normal object layout computation where the the members are padded for representability and then overaligned for representability, and it would probably be easier to understand and less messy if you pretend to be a compiler doing struct layout. Something like:
```
size_t length = 0;

size_t call_context_length = __builtin_cheri_round_representable_length(sizeof(struct call_context));
length = __align_up(length, _Alignof(struct call_context));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(call_context_length) + 1);
size_t call_context_offset = length;
length += call_context_length;

size_t stack_bytes_length = __builtin_cheri_round_representable_length(stack_bytes);
length = __align_up(length, _Alignof(max_align_t));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(stack_bytes_length) + 1);
size_t stack_bytes_offset = length;
length += stack_bytes_length;

size_t call_frame_length = __builtin_cheri_round_representable_length(sizeof(struct call_frame));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(call_frame_length) + 1);
size_t call_frame_offset = length;
length += call_frame_length;

size_t rsize_length = __builtin_cheri_round_representable_length(rsize);
length = __align_up(length, _Alignof(max_align_t));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(rsize_length) + 1);
size_t rsize_offset = length;
length += rsize_length;
```

Then you just `alloca(length)`.

Or with a macro for the repetition:

```
#define LAYOUT_SUBOBJECT(name, size, align) \
    size_t name##_length = __builtin_cheri_round_representable_length(size); \
    length = __align_up(length, align); \
    length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(name##_length) + 1); \
    size_t name##_offset = length; \
    length += name##_length

LAYOUT_SUBOBJECT(call_context, sizeof(struct call_context), _Alignof(struct call_context));
LAYOUT_SUBOBJECT(stack_bytes, stack_bytes, _Alignof(max_align_t));
LAYOUT_SUBOBJECT(call_frame, sizeof(struct call_frame), _Alignof(struct call_frame));
LAYOUT_SUBOBJECT(rsize, rsize, _Alignof(max_align_t));
```

A lot of that should be able to be constant folded, although Morello LLVM is missing a bunch of that logic in its backend so perhaps not. If you indirect the builtins via stubs that are the identity and constant 1 respectively that code also works just fine on non-CHERI so you don't need to #ifdef it. You may want to adapt the macro to take the actual variable names and declare them at the start of the function though:

```
#define LAYOUT_SUBOBJECT(subobj_length, subobj_offset, size, align) \
    subobj_length = __builtin_cheri_round_representable_length(size); \
    length = __align_up(length, align); \
    length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(subobj_length) + 1); \
    subobj_offset = length; \
    length += subobj_length

LAYOUT_SUBOBJECT(call_context_length, call_context_offset, sizeof(struct call_context), _Alignof(struct call_context));
LAYOUT_SUBOBJECT(stack_bytes_length, stack_bytes_offset, stack_bytes, _Alignof(max_align_t));
LAYOUT_SUBOBJECT(call_frame_length, call_frame_offset, sizeof(struct call_frame), _Alignof(struct call_frame));
LAYOUT_SUBOBJECT(rsize_length, rsize_offset, rsize, _Alignof(max_align_t));
```

Part of the benefit of this is that you only do the representability rounding in one place, rather than the current approach where you compute it once to figure out how much space to allocate and then re-round things as you go with the resulting pointer, hoping that the calculations match up. By fully computing the layout up-front and reusing the same values from the calculation you know it'll match up, so long as the original caluclation was correct, which is easy to verify, especially with the macro.

Reply:
The reason we re-round things as we go is because we worked on the assumption that the pointer returned by `alloca` might be as little as 16-byte aligned, which, on reflection, is clearly too weak.

Reply:
It'll be as aligned as the entire allocation needs to be representable, which will be at least as aligned as any sub-object needs for representability

Reply:
Thank you! This is way cleaner. I've changed it a bit to call `_Alignof()` in the macro, and I've introduced additional macros not to ifdef code as you suggested.

I've added you as a co-author of this change.